Add bidirectional packet capture

AryanBakliwal opened the pull request:
fixes: #6862 

Added `bidirectional` field in packet capture CR spec.
For testing, I created two pods and pinged one from the other. 

Screenshot of the `.pcapng` output file.
![image](https://github.com/user-attachments/assets/34262988-987f-43ba-870c-ca82cc02089d)

AryanBakliwal commented:
@hangyan please let me know what you think about this

antoninbas commented on pkg/agent/packetcapture/packetcapture_controller.go:
```diff
@@ -464,10 +464,22 @@ func (c *Controller) performCapture(
+
+	var packets chan gopacket.Packet
+	if pc.Spec.Bidirection {
+		reqpackets, err1 := c.captureInterface.Capture(ctx, device, snapLen, srcIP, dstIP, pc.Spec.Packet)
+		respackets, err2 := c.captureInterface.Capture(ctx, device, snapLen, dstIP, srcIP, pc.Spec.Packet)
```
@hangyan do we need 2 separate capture instances, or a single instance that can capture traffic in both directions?

hangyan replied:
one is enough. The bpf code should be updated to support bidirection capture.

@AryanBakliwal You can use tcpdump to check the bpf coded it generated and update ours corresponding

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -439,6 +439,8 @@ type PacketCaptureSpec struct {
 	// for a capture session, and at least one `Pod` should be present either in the source or the destination.
 	Source      Source      `json:"source"`
 	Destination Destination `json:"destination"`
+	// Bidirection specifies whether to capture the return (response) traffic from the destination back to the source.
+	Bidirection bool `json:"bidirection"`
```
maybe we could have an "enum" with such values: `SourceToDestination`, `DestinationToSource`, `Both` (and name the field `direction`).

What do you think?
cc @tnqn @hongliangl

tnqn replied:
Sounds good to me

AryanBakliwal replied:
Got it, I’ll make the required changes

AryanBakliwal commented:
Hey @hangyan @antoninbas, I made the changes in the bpf code according to the one generated by tcpdump. 

When I try to test the bidirectional packet capture, it fails with this log
```
E1229 11:44:58.723258       1 packetcapture_controller.go:424] "PacketCapture failed capturing packets" err="invalid argument" name="pc-test"
```

Could you please take a look and help me identify what might be going wrong?

Also, golangci-lint is giving me this error even though I have changed the `Capture` method everywhere.
```
===> Running golangci (linux) <===
===> Running golangci (windows) <===
cmd/antrea-agent/agent.go:64:2: could not import antrea.io/antrea/pkg/agent/packetcapture (-: # antrea.io/antrea/pkg/agent/packetcapture
pkg/agent/packetcapture/packetcapture_controller.go:170:23: cannot use capture (variable of type *capture.pcapCapture) as PacketCapturer value in assignment: *capture.pcapCapture does not implement PacketCapturer (wrong type for method Capture)
                have Capture(context.Context, string, int, "net".IP, "net".IP, *"antrea.io/antrea/pkg/apis/crd/v1alpha1".Packet) (chan gopacket.Packet, error)
                want Capture(context.Context, string, int, "net".IP, "net".IP, *"antrea.io/antrea/pkg/apis/crd/v1alpha1".Packet, bool) (chan gopacket.Packet, error)) (typecheck)
        "antrea.io/antrea/pkg/agent/packetcapture"
        ^
pkg/agent/packetcapture/capture_interface.go:1: : # antrea.io/antrea/pkg/agent/packetcapture [antrea.io/antrea/pkg/agent/packetcapture.test]
pkg/agent/packetcapture/packetcapture_controller.go:170:23: cannot use capture (variable of type *capture.pcapCapture) as PacketCapturer value in assignment: *capture.pcapCapture does not implement PacketCapturer (wrong type for method Capture)
                have Capture(context.Context, string, int, "net".IP, "net".IP, *"antrea.io/antrea/pkg/apis/crd/v1alpha1".Packet) (chan gopacket.Packet, error)
                want Capture(context.Context, string, int, "net".IP, "net".IP, *"antrea.io/antrea/pkg/apis/crd/v1alpha1".Packet, bool) (chan gopacket.Packet, error) (typecheck)
// Copyright 2024 Antrea Authors.
make: *** [Makefile:329: golangci] Error 1
```

hangyan commented:
> Hey @hangyan @antoninbas, I made the changes in the bpf code according to the one generated by tcpdump.
> 
> When I try to test the bidirectional packet capture, it fails with this log
> 
> ```
> E1229 11:44:58.723258       1 packetcapture_controller.go:424] "PacketCapture failed capturing packets" err="invalid argument" name="pc-test"
> ```
> 
> Could you please take a look and help me identify what might be going wrong?
> 
> Also, golangci-lint is giving me this error even though I have changed the `Capture` method everywhere.
> 
> ```
> ===> Running golangci (linux) <===
> ===> Running golangci (windows) <===
> cmd/antrea-agent/agent.go:64:2: could not import antrea.io/antrea/pkg/agent/packetcapture (-: # antrea.io/antrea/pkg/agent/packetcapture
> pkg/agent/packetcapture/packetcapture_controller.go:170:23: cannot use capture (variable of type *capture.pcapCapture) as PacketCapturer value in assignment: *capture.pcapCapture does not implement PacketCapturer (wrong type for method Capture)
>                 have Capture(context.Context, string, int, "net".IP, "net".IP, *"antrea.io/antrea/pkg/apis/crd/v1alpha1".Packet) (chan gopacket.Packet, error)
>                 want Capture(context.Context, string, int, "net".IP, "net".IP, *"antrea.io/antrea/pkg/apis/crd/v1alpha1".Packet, bool) (chan gopacket.Packet, error)) (typecheck)
>         "antrea.io/antrea/pkg/agent/packetcapture"
>         ^
> pkg/agent/packetcapture/capture_interface.go:1: : # antrea.io/antrea/pkg/agent/packetcapture [antrea.io/antrea/pkg/agent/packetcapture.test]
> pkg/agent/packetcapture/packetcapture_controller.go:170:23: cannot use capture (variable of type *capture.pcapCapture) as PacketCapturer value in assignment: *capture.pcapCapture does not implement PacketCapturer (wrong type for method Capture)
>                 have Capture(context.Context, string, int, "net".IP, "net".IP, *"antrea.io/antrea/pkg/apis/crd/v1alpha1".Packet) (chan gopacket.Packet, error)
>                 want Capture(context.Context, string, int, "net".IP, "net".IP, *"antrea.io/antrea/pkg/apis/crd/v1alpha1".Packet, bool) (chan gopacket.Packet, error) (typecheck)
> // Copyright 2024 Antrea Authors.
> make: *** [Makefile:329: golangci] Error 1
> ```

this error was reported by golangci on windows, we have a `pcap_unsupported.go` for windows/... interface, it's an non-op since we didn't support this on windows yet, you should change that too.

hangyan commented on pkg/agent/packetcapture/capture/bpf.go:
```diff
@@ -214,6 +275,10 @@ func calculateInstructionsSize(packet *crdv1alpha1.Packet) int {
 	// src and dst ip
 	count += 4
 
+	if bidirection {
```
this part seems incorrect. If we didn't filter on ports and only on IP, only 4 extra instructions is needed. So it's better to break it down. I'm guessing the error is mostly like caused by the invalid jump distance.

AryanBakliwal replied:
Thanks for pointing this out, after breaking it and doing some more changes in `compilePacketFilter`, it now works as expected.

AryanBakliwal commented:
I’ve added the `direction` field with enum values (`SourceToDestination`, `DestinationToSource`, and `Both`) and set the default to `SourceToDestination`. Additionally, I’ve updated the YAML manifests, packet capture guide and tests accordingly.

Please let me know if any adjustments are needed or if there's anything else I should update or add.

### Captured packets
ICMP (SourceToDestination)
![Screenshot from 2025-01-06 00-07-02](https://github.com/user-attachments/assets/4c0f6253-6bf8-486b-b731-96f98d6db5f1)

ICMP (DestinationToSource)
![Screenshot from 2025-01-06 00-14-28](https://github.com/user-attachments/assets/6dae03cc-4679-4afa-ad78-81ec5207eb0d)

ICMP (Both)
![Screenshot from 2025-01-06 00-17-58](https://github.com/user-attachments/assets/9de5934f-55a3-401e-93b7-30d7dd144295)

TCP (SourceToDestination)
![image](https://github.com/user-attachments/assets/cfd1e30d-5112-4f7f-84f0-4a08cd028b80)

TCP (DestinationToSource)
![image](https://github.com/user-attachments/assets/d625a4a1-6c09-443c-998d-1b2c1999de82)

TCP (Both)
![image](https://github.com/user-attachments/assets/2cad9977-fa85-4e7b-b6db-74a8618cf8e1)

hangyan commented on pkg/agent/packetcapture/capture/bpf.go:
```diff
@@ -134,18 +122,96 @@ func compilePacketFilter(packetSpec *crdv1alpha1.Packet, srcIP, dstIP net.IP) []
 		}
 	}
 
+	// source ip
```
In the future we may also support tcp flags and other layer4 configs, if that happens, we should consider make the current code structure more modularized, or it would be extremely hard to extend this function .  This won't be easy but i suggest to review this part and see if we can do better. 

not sure if we can separate the ip section and ports section apart, call  sub functions to calculate their instruments size and sums up.

AryanBakliwal replied:
I'll try my best to improve the code structure and explore separating the IP and port sections as suggested

AryanBakliwal replied:
I am thinking of refactoring the code as you suggested by modularizing the logic for IP and port compilation.
I’ve created a function `compareIP`  to handle the conditional logic for building the instruction to compare IP address depending on the direction and which IP address is loaded for comparison (the boolean `isSrc` tells if I have loaded the source or destination IP address)

```
func compareIP(direction crdv1alpha1.CaptureDirection, skipTrue, skipFalse uint8, isSrc bool, srcAddrVal, dstAddrVal uint32) bpf.Instruction {
	if isSrc {
		if direction == crdv1alpha1.CaptureDirectionDestinationToSource {
			return bpf.JumpIf{Cond: bpf.JumpEqual, Val: dstAddrVal, SkipTrue: skipTrue, SkipFalse: skipFalse}
		} else {
			return bpf.JumpIf{Cond: bpf.JumpEqual, Val: srcAddrVal, SkipTrue: skipTrue, SkipFalse: skipFalse}
		}
	} else {
		if direction == crdv1alpha1.CaptureDirectionDestinationToSource {
			return bpf.JumpIf{Cond: bpf.JumpEqual, Val: srcAddrVal, SkipTrue: skipTrue, SkipFalse: skipFalse}
		} else {
			return bpf.JumpIf{Cond: bpf.JumpEqual, Val: dstAddrVal, SkipTrue: skipTrue, SkipFalse: skipFalse}
		}
	}
}
```
Similarly a function `comparePort` builds instruction for comparing port
Other functions `loadSrcPort` and `loadDstPort` checks if to load ports for comparison depending upon the direction and port value from the packetSpec
Additional functions to calculate number of `SkipTrue` and `SkipFalse` instructions dynamically

Let me know if this looks better or if you have any suggestions for further improvements.

hangyan commented on pkg/agent/packetcapture/capture/bpf_test.go:
```diff
@@ -173,11 +202,91 @@ func TestPacketCaptureCompileBPF(t *testing.T) {
+		},
+		{
+			name:  "with-proto-port-and-Both",
+			srcIP: net.ParseIP("127.0.0.1"),
```
better to add a test case for only have `dstPort` and `Both`, that's the real world case as client port is always random.

AryanBakliwal replied:
makes sense, will add this test case

hangyan commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -439,6 +447,9 @@ type PacketCaptureSpec struct {
 	// for a capture session, and at least one `Pod` should be present either in the source or the destination.
 	Source      Source      `json:"source"`
 	Destination Destination `json:"destination"`
+	// Direction specifies which packets to capture (source -> destination, destination -> source or both).
```
you may need to run `make codegen`

AryanBakliwal replied:
I did ran `make codegen` but it did not make any changes in the `zz_generated.deepcopy.go` file, maybe because it is not a pointer in `PacketCaptureSpec`.

antoninbas replied:
Yes, I think @AryanBakliwal is correct here. A shallow-copy is sufficient for a string, and the `*out = *in` statement in `func (in *PacketCaptureSpec) DeepCopyInto(out *PacketCaptureSpec)` will take care of copying the field value.

The CI job that checks for code generation is also passing right now.

AryanBakliwal commented on pkg/agent/packetcapture/capture/bpf.go:
```diff
@@ -178,7 +244,38 @@ func compilePacketFilter(packetSpec *crdv1alpha1.Packet, srcIP, dstIP net.IP) []
+// (006) ld       [30]									   # Load 4B at 30 (dest address)
+// (007) jeq      #0xaf40103       jt 8	jf 26			   # If bytes match(10.244.0.3), goto #8, else #26
+// (008) ldh      [20]									   # Load 2B at 20 (13b Fragment Offset)
+// (009) jset     #0x1fff          jt 26	jf 10	   # Use 0x1fff as a mask for fragment offset; If fragment offset != 0, #10, else #26
```
sure

antoninbas commented on pkg/agent/packetcapture/capture/bpf.go:
```diff
@@ -101,20 +101,8 @@ func compilePacketFilter(packetSpec *crdv1alpha1.Packet, srcIP, dstIP net.IP) []
-		addrVal := binary.BigEndian.Uint32(dstIP[len(dstIP)-4:])
-		inst = append(inst, bpf.JumpIf{Cond: bpf.JumpEqual, Val: addrVal, SkipTrue: 0, SkipFalse: size - uint8(len(inst)) - 2})
-	}
+	srcAddrVal := binary.BigEndian.Uint32(srcIP[len(srcIP)-4:])
+	dstAddrVal := binary.BigEndian.Uint32(dstIP[len(dstIP)-4:])
```
I am not sure I understand this change. Isn't it possible for `srcIP` and `dstIP` to be nil?

hangyan replied:
I think the current spec enforced `srcIP` and `dstIP` both cannot be nil, because either Pod or IP will present.  It cloud be pretty easy to remove the restrictions in the crd spec and do a little tweak to make this work.(no the target of the PR)

antoninbas commented on pkg/apis/crd/v1alpha1/types.go:
```diff
@@ -431,6 +431,14 @@ type PacketCaptureFileServer struct {
 
+type CaptureDirection string
+
+const (
+	SourceToDestination CaptureDirection = "SourceToDestination"
+	DestinationToSource CaptureDirection = "DestinationToSource"
+	Both                CaptureDirection = "Both"
+)
```
```suggestion
const (
	CaptureDirectionSourceToDestination CaptureDirection = "SourceToDestination"
	CaptureDirectionDestinationToSource CaptureDirection = "DestinationToSource"
	CaptureDirectionBoth                CaptureDirection = "Both"
)
```

antoninbas commented on build/yamls/antrea.yml:
```diff
@@ -3035,7 +3035,10 @@ spec:
                               minimum: 1
                               maximum: 65535
-
+                direction:
+                  type: string
+                  enum: ["SourceToDestination", "DestinationToSource", "Both"]
+                  default: "SourceToDestination"
```
you seem to have made that change manually, which is not correct
the source of truth is `build/charts/antrea/crds/packetcapture.yaml`. You need to edit that file and run `make manifest`.

AryanBakliwal replied:
Thanks for the clarification, I'll make the changes in the correct files and regenerate the manifests.

antoninbas reviewed (commented):
could we add an e2e test with the `Both` direction?

AryanBakliwal commented:
@antoninbas @hangyan I have refactored the `compilePacketFilter` function, added a test case for `dstPort` and `Both` in `bpf_test.go`, generated the manifests by running `make manifest` and added an e2e test with the `Both` direction. 
Could you please review the changes? Let me know if there’s anything else I can improve.

antoninbas commented on pkg/agent/packetcapture/capture/bpf.go:
```diff
@@ -214,6 +332,31 @@ func calculateInstructionsSize(packet *crdv1alpha1.Packet) int {
+	if direction == crdv1alpha1.CaptureDirectionBoth {
+		count += 3
+
+		transPort := packet.TransportHeader
+		if transPort.TCP != nil {
+			// load Fragment Offset
+			count += 3
+			if transPort.TCP.SrcPort != nil {
+				count += 2
+			}
+			if transPort.TCP.DstPort != nil {
+				count += 2
+			}
+
+		} else if transPort.UDP != nil {
+			count += 3
+			if transPort.UDP.SrcPort != nil {
+				count += 2
+			}
+			if transPort.UDP.DstPort != nil {
+				count += 2
+			}
+		}
```
correct me if I'm wrong but this code looks like a perfect duplicate of the code above
you could do something like this:
```go
transport := packet.TransportHeader
portFiltersSize := func() int {
                count := 0
		if transport.TCP != nil {
			// load Fragment Offset
			count += 3
			if transport.TCP.SrcPort != nil {
				count += 2
			}
			if transport.TCP.DstPort != nil {
				count += 2
			}

		} else if transport.UDP != nil {
			count += 3
			if transport.UDP.SrcPort != nil {
				count += 2
			}
			if transport.UDP.DstPort != nil {
				count += 2
			}
		}
		return count
}
```

Note that I have changed the spelling of `transPort` to `transport`, as the current version makes no sense.

AryanBakliwal replied:
You're correct, it is indeed a copy. I'll replace it with this function to remove duplication.
Thanks for sharing this.

antoninbas commented on pkg/agent/packetcapture/capture/bpf.go:
```diff
@@ -214,6 +332,31 @@ func calculateInstructionsSize(packet *crdv1alpha1.Packet) int {
 	count += 4
 
+	if direction == crdv1alpha1.CaptureDirectionBoth {
+		count += 3
```
can you add a comment for this one?

based on your function comment, in the `Both` direction case, we have to repeat the source host and destination host filters. Is that what this is for? This seems "small" to me, as I see the following above:
```
	// src and dst ip
	count += 4
```
but I assume you validated it in your unit tests

AryanBakliwal replied:
Yes, this is for the IP check instructions. We only need 3 more in the `Both` case because the source IP address is already loaded and 
```
// (004) ld       [26]                                     # Load 4B at 26 (source address)
```
we jump to compare it with the destination IP address.

I will add a comment for this.

antoninbas commented on pkg/agent/packetcapture/capture/bpf.go:
```diff
@@ -214,6 +332,31 @@ func calculateInstructionsSize(packet *crdv1alpha1.Packet) int {
+	if direction == crdv1alpha1.CaptureDirectionBoth {
+		count += 3
+
+		transPort := packet.TransportHeader
```
I notice that you are not checking for `packet != nil` here, which doesn't match the above code?

AryanBakliwal replied:
Thanks for pointing this out, I overlooked it. Will add this check.

antoninbas commented on pkg/agent/packetcapture/capture/bpf.go:
```diff
@@ -134,18 +178,61 @@ func compilePacketFilter(packetSpec *crdv1alpha1.Packet, srcIP, dstIP net.IP) []
 	}
 
+	// source ip
+	if srcIP != nil {
+		inst = append(inst, loadIPv4SourceAddress)
+		// from here we need to check the inst length to calculate skipFalse. If no protocol is set, there will be no related bpf instructions.
+		inst = append(inst, compareIP(direction, 0, calculateSkipFalse(size, uint8(len(inst)), srcPort, dstPort, direction), true, srcAddrVal, dstAddrVal))
+	}
+	// dst ip
+	if dstIP != nil {
```
the code above assumes that `srcIP` and `dstIP` are never nil. There was a discussion about it, which was marked as resolved. We should be consistent. If this is a precondition for the function, then we should not include these nil checks.

AryanBakliwal replied:
You're right, I will remove these nil checks.

antoninbas commented on pkg/agent/packetcapture/capture/bpf.go:
```diff
@@ -72,11 +72,67 @@ func compareProtocol(protocol uint32, skipTrue, skipFalse uint8) bpf.Instruction
 	return bpf.JumpIf{Cond: bpf.JumpEqual, Val: protocol, SkipTrue: skipTrue, SkipFalse: skipFalse}
 }
 
+func calculateSkipTrue(size, instlen uint8, direction crdv1alpha1.CaptureDirection, srcPort, dstPort uint16, isSrc bool) uint8 {
```
the way these helper functions are written doesn't make a lot of sense to me. The functions have a generic name, yet they have very specialized implementation that is dependent on a bunch of parameters. To me, the fact that `direction` is a parameter to this function is a bit of a red flag.

overall, I would expect the implementation to be more streamlined. There should be a helper function to generate IP + port matching. We can call the function for the `SourceToDestination` direction. We can also call the function for the `DestinationToSource` direction by swapping src / dst function parameters. For the `Both` direction, the function should just be called twice, and the caller should add the glue to make it work (`or` operator). Am I missing something? Does it have to be much more complex than that?

AryanBakliwal replied:
I have refactored the code to contain the logic to compare IP and ports in a single function that can be called by interchanging function parameters for `SourceToDestination` and `DestinationToSource`; and called twice for `Both`.

antoninbas reviewed (commented):
thanks for your ongoing work on this

antoninbas commented on pkg/agent/packetcapture/capture/bpf.go:
```diff
@@ -169,16 +256,47 @@ func compilePacketFilter(packetSpec *crdv1alpha1.Packet, srcIP, dstIP net.IP) []
+// (004) ld       [26]									   # Load 4B at 26 (source address)
+// (005) jeq      #0xaf40102       jt 6	jf 15			   # If bytes match(10.244.0.2), goto #6, else #15
+// (006) ld       [30]									   # Load 4B at 30 (dest address)
+// (007) jeq      #0xaf40103       jt 8	jf 26			   # If bytes match(10.244.0.3), goto #8, else #26
```
nit: The IP addresses don't match the hex value and the human-readable filter that you have above (should be 10.244.1.2 and 10.244.1.3)

AryanBakliwal replied:
Thanks for catching that.

antoninbas commented on pkg/agent/packetcapture/capture/bpf.go:
```diff
@@ -169,16 +256,47 @@ func compilePacketFilter(packetSpec *crdv1alpha1.Packet, srcIP, dstIP net.IP) []
+// (004) ld       [26]									   # Load 4B at 26 (source address)
+// (005) jeq      #0xaf40102       jt 6	jf 15			   # If bytes match(10.244.0.2), goto #6, else #15
+// (006) ld       [30]									   # Load 4B at 30 (dest address)
+// (007) jeq      #0xaf40103       jt 8	jf 26			   # If bytes match(10.244.0.3), goto #8, else #26
```
would the code be easier to generate if we jumped to (015) in case this instruction (007) was not a match?

the fact that we know we can skip the second `or` operand in this case is an optimization. Not saying it's not a good optimization, but if using a "fixed" offset makes code generation much easier, maybe we should do it for now.

I'm thinking about the discussion we just had during the community meeting. That being said, I can see why generating the same code as tcpdump is valuable.

antoninbas replied:
@AryanBakliwal Please take a look at this ^

Taking a concrete example:
```
tcpdump -d "ip proto 6 and ((src host 10.244.1.2 and dst host 10.244.1.3) or (src host 10.244.1.3 and dst host 10.244.1.2))"
```
This is the code generated by tcpdump:
```
(000) ldh      [12]
(001) jeq      #0x800           jt 2	jf 12
(002) ldb      [23]
(003) jeq      #0x6             jt 4	jf 12
(004) ld       [26]
(005) jeq      #0xaf40102       jt 6	jf 8
(006) ld       [30]
(007) jeq      #0xaf40103       jt 11	jf 12
(008) jeq      #0xaf40103       jt 9	jf 12
(009) ld       [30]
(010) jeq      #0xaf40102       jt 11	jf 12
(011) ret      #262144
(012) ret      #0
```

If we generate this code instead, I believe it will be simpler:
```
(000) ldh      [12]
(001) jeq      #0x800           jt 2	jf 12
(002) ldb      [23]
(003) jeq      #0x6             jt 4	jf 12
(004) ld       [26]
(005) jeq      #0xaf40102       jt 6	jf 8
(006) ld       [30]
(007) jeq      #0xaf40103       jt 11	jf 8   <- less efficient than tcpdump's version but easier to implement?
(008) jeq      #0xaf40103       jt 9	jf 12
(009) ld       [30]
(010) jeq      #0xaf40102       jt 11	jf 12
(011) ret      #262144
(012) ret      #0
```

AryanBakliwal replied:
@antoninbas Actually, the real problem is the calculation of `skipTrue` (for which I created the function `calculateSkipTrue`) when we want to accept the packet. But `calculateSkipTrue` depends on many parameters [`srcPort`, `dstPort`, `direction`, which port is loaded in the register (`srcPort` or `dstPort`) and if we are calling this function from the first half (instructions that check request) or the second half (instructions that check response)] to calculate the number of instructions to skip to jump to the `returnKeep` instruction. This makes the code generation very complex if generating instruction set exactly like tcpdump.

Taking each case:

1. ICMP (both `srcPort` and `dstPort` are 0)
```
tcpdump -d "ip proto 6 and ((src host 10.244.1.2 and dst host 10.244.1.3) or (src host 10.244.1.3 and dst host 10.244.1.2))"
```
Code generated by tcpdump:
```
(000) ldh      [12]
(001) jeq      #0x800           jt 2	jf 12
(002) ldb      [23]
(003) jeq      #0x6             jt 4	jf 12
(004) ld       [26]
(005) jeq      #0xaf40102       jt 6	jf 8
(006) ld       [30]
(007) jeq      #0xaf40103       jt 11	jf 12    <- need to jump to 11 from 7, difference to calculate = 3
(008) jeq      #0xaf40103       jt 9	jf 12
(009) ld       [30]
(010) jeq      #0xaf40102       jt 11	jf 12    <- skipTrue = 0
(011) ret      #262144
(012) ret      #0
```

2. TCP (`srcPort` = 0 and `dstPort` > 0)
```
tcpdump -d "ip proto 6 and ((src host 10.244.1.2 and dst host 10.244.1.3 and dst port 80) or (src host 10.244.1.3 and dst host 10.244.1.2 and src port 80))"
```
Code generated by tcpdump:
```
(000) ldh      [12]
(001) jeq      #0x800           jt 2	jf 22
(002) ldb      [23]
(003) jeq      #0x6             jt 4	jf 22
(004) ld       [26]
(005) jeq      #0xaf40102       jt 6	jf 13
(006) ld       [30]
(007) jeq      #0xaf40103       jt 8	jf 22    <- jump to next instruction since port(s) not zero
(008) ldh      [20]
(009) jset     #0x1fff          jt 22	jf 10
(010) ldxb     4*([14]&0xf)
(011) ldh      [x + 16]
(012) jeq      #0x50            jt 21	jf 22    <- need to jump to 21 from 12, difference to calculate = 8
(013) jeq      #0xaf40103       jt 14	jf 22
(014) ld       [30]
(015) jeq      #0xaf40102       jt 16	jf 22
(016) ldh      [20]
(017) jset     #0x1fff          jt 22	jf 18
(018) ldxb     4*([14]&0xf)
(019) ldh      [x + 14]
(020) jeq      #0x50            jt 21	jf 22    <- skipTrue = 0
(021) ret      #262144
(022) ret      #0								 
```
Note that this is similar to when (`srcPort` > 0 and `dstPort` = 0) and (`srcPort` > 0 and `dstPort` > 0).

Since the problem is to jump to the `returnKeep` instruction, why not add it again to mitigate this complex calculation and jump to it easily. For example:

1. ICMP:
Code that can be generated more simply
```
(000) ldh      [12]
(001) jeq      #0x800           jt 2	jf 12
(002) ldb      [23]
(003) jeq      #0x6             jt 4	jf 12
(004) ld       [26]
(005) jeq      #0xaf40102       jt 6	jf 8
(006) ld       [30]
(007) jeq      #0xaf40103       jt 8	jf 13    <- just skip to next instruction if true
(008) ret      #262144
(009) jeq      #0xaf40103       jt 10	jf 13
(010) ld       [30]
(011) jeq      #0xaf40102       jt 12	jf 13
(012) ret      #262144
(013) ret      #0
```

2. TCP (`srcPort` = 0 and `dstPort` > 0)
Code that can be generated more simply
```
(000) ldh      [12]
(001) jeq      #0x800           jt 2	jf 23
(002) ldb      [23]
(003) jeq      #0x6             jt 4	jf 23
(004) ld       [26]
(005) jeq      #0xaf40102       jt 6	jf 13
(006) ld       [30]
(007) jeq      #0xaf40103       jt 8	jf 23
(008) ldh      [20]
(009) jset     #0x1fff          jt 23	jf 10
(010) ldxb     4*([14]&0xf)
(011) ldh      [x + 16]
(012) jeq      #0x50            jt 13	jf 23    <- just skip to next instruction if true
(013) ret      #262144
(014) jeq      #0xaf40103       jt 15	jf 23
(015) ld       [30]
(016) jeq      #0xaf40102       jt 17	jf 23
(017) ldh      [20]
(018) jset     #0x1fff          jt 23	jf 19
(019) ldxb     4*([14]&0xf)
(020) ldh      [x + 14]
(021) jeq      #0x50            jt 22	jf 23
(022) ret      #262144
(023) ret      #0								 
```

Let me know what you think about this approach

antoninbas replied:
Adding an extra return instruction sounds fine to me.

I think what we are going for is some code that looks like this in the end:
```go
if direction == SourceToDest {
        generateFilters(offset, srcIP, dstIP, srcPort, dstPort)
} else if direction == DestToSource {
        generateFilters(offset, dstIP, srcIP, dstPort, srcPort)
} else {
        generateFilters(offset, srcIP, dstIP, srcPort, dstPort)
        // ...
        generateFilters(offset, dstIP, srcIP, dstPort, srcPort)
}
```
So if we can get there and have some generate code that looks reasonable and works, that would be good.

AryanBakliwal replied:
@antoninbas I have made the requested changes and refactored the code as you suggested and as per our discussion. PTAL and let me know if there is anything else to update or add.

antoninbas commented on pkg/agent/packetcapture/capture/bpf.go:
```diff
@@ -169,50 +210,131 @@ func compilePacketFilter(packetSpec *crdv1alpha1.Packet, srcIP, dstIP net.IP) []
+// (012) jeq      #0x7b            jt 13	jf 27		   # TCP Source Port: If 123, goto #13, else #26
+// (013) ldh      [x + 16]								   # Load 2B at x+16 (TCP dst port)
+// (014) jeq      #0x7c            jt 15	jf 27		   # TCP dst port: If 123, goto #15, else #26
+// (015) ret      #262144								   # MATCH
```
Maybe a stupid question, but if this is the only change we have, it doesn't seem worth it? Why can't we calculate the jump size using `size - curOffset - 1` given that:
1) we always compute the full size up front
2) we always know the current offset ("where we are")
3) the instruction before last is always a `MATCH` return instruction 

What am I missing?

AryanBakliwal replied:
@antoninbas You're right, the jump size can indeed be calculated easily with the available information (full size and current offset). 
The main challenge, however, isn't the calculation itself but rather when to perform it, meaning in which instructions do we need `skipTrue` set to `size - curOffset - 1`. 
And if a helper function is used for calculation, then it depends on which instruction is calling the helper function, what is the direction and what is the value of ports. Adding this extra instruction helps simplify the instruction set generation and avoids unnecessary complexity.

### The challenges I faced:
For request packets in `Both` direction, there are 3 cases (assuming the 3 instructions - dstIP. srcPort and srcPort comparison instructions are calling a helper function to get jump size/`skipTrue` value):
- When both ports are 0, the destination IP comparison instruction performs the calculation to jump to `MATCH`.
- When only one port is 0, the destination IP comparison sets skipTrue to 0, and the comparison for the defined port performs the calculation to jump to `MATCH`.
- When neither port is 0, all comparisons set skipTrue to 0, except for the destination port comparison, which performs the calculation to jump to `MATCH`. 

In `SourceToDestination`, `DestinationToSource` and response packets in `Both`, no calculations are needed, and these 3 instructions (dstIP, srcPort and dstPort comparison) jump to `MATCH` step with `skipTrue` set to 0. 

The helper function that calculates the jump size or returns the value of `skipTrue` requires additional context which is a bunch of parameters such as the calling instruction, direction, ports, size and current offset. Adding this logic in the code makes it difficult to extend in the future if we filter traffic on more parameters and also increases complexity.

Hope that this explanation makes sense and adding an extra `MATCH` instruction seems reasonable. Any suggestions on how we could simplify or approach this differently? I'm open to any ideas you might have for improving the implementation of calculation of `skipTrue` value.

hangyan commented on pkg/agent/packetcapture/capture/bpf.go:
```diff
@@ -72,11 +72,67 @@ func compareProtocol(protocol uint32, skipTrue, skipFalse uint8) bpf.Instruction
+	return count
+}
+
+func compareIPPort(srcAddrVal, dstAddrVal uint32, size, curLen uint8, srcPort, dstPort uint16, useSkipFalse bool) []bpf.Instruction {
```
this function name is not very straightforward, can you use a new one? we may also need some comments.

Also i think the param `useSkipFalse` maybe misleading, we always `use skip false`, the difference is skipFalse to ret `#0` or skipFalse to some intermediate instructions.

AryanBakliwal replied:
@hangyan For function name, I think `generateFilter` or `generateInstructions` can be used. For the bool param, we can use something like `jumpToResponseCheck` or `skipRequestCheck`?

hangyan replied:
`compilePacketFilter` call `compareIPPort`, so `compileIPPortFilter` or `genereateIPPortFilterInstructions` should be fine. (compile not compare)

For `useSkipFalse`, seems either one need detailed comment for  documentation.

AryanBakliwal replied:
I will change the function and param name, and also add a detailed comment for it.

AryanBakliwal commented on pkg/agent/packetcapture/capture/bpf.go:
```diff
@@ -169,50 +210,131 @@ func compilePacketFilter(packetSpec *crdv1alpha1.Packet, srcIP, dstIP net.IP) []
+// 'ip proto 6 and ((src host 10.244.1.2 and dst host 10.244.1.3 and src port 123 and dst port 124) or (src host 10.244.1.3 and dst host 10.244.1.2 and src port 124 and dst port 123))'
+// And using `tcpdump -i <device> '<filter>' -d` will generate the following BPF instructions:
+// (000) ldh      [12]									   # Load 2B at 12 (Ethertype)
+// (001) jeq      #0x800           jt 2	jf 26			   # Ethertype: If IPv4, goto #2, else #26
```
This is the example instruction set generated by tcpdump in which we skip to 26. Example of our implementation is below, in which we jump to 27.

hangyan replied:
ok, can you change the comment in that part to `#27`,  or the comment will be misleading.

AryanBakliwal replied:
I'll change this and also check if any other comments need correction

AryanBakliwal commented:
@antoninbas @hangyan Changed the function and param name, and added some comments. Please let me know if there’s anything else I should take care of.

hangyan reviewed (approved):
Thanks for your work

luolanzone commented:
Hi @AryanBakliwal please re-base to the latest main branch and squash all your changes into one commit. Thanks.

antoninbas commented on pkg/agent/packetcapture/capture/bpf.go:
```diff
@@ -72,11 +72,74 @@ func compareProtocol(protocol uint32, skipTrue, skipFalse uint8) bpf.Instruction
+	// we need to check if the packet is a response (from destination to source). In this case, we skip to the instruction where we compare the 
+	// loaded source IP with the destination IP from the packet spec. The skipRequestCheck flag indicates whether we need to call calculateSkipFalse 
+	// to determine how many instructions to skip before checking for response packets or just skip to the last instruction (drop packet).
+	if skipRequestCheck {
```
is there a better name than `skipRequestCheck`? Even after reading the comment, the name still doesn't make sense to me.

AryanBakliwal replied:
@antoninbas Since it tells if we need to jump to return traffic check part of the filter, how about `checkIfResponse` or `jumpToResponseCheck` ?

antoninbas replied:
IMO the current variable name is the "opposite" of what it should be almost.
Maybe I would go with `needsOtherTrafficDirectionCheck` as a drop-in replacement for `skipRequestCheck`?

AryanBakliwal replied:
`needsOtherTrafficDirectionCheck` makes sense. Or maybe `needsReturnTrafficDirectionCheck` to be more specific?

antoninbas replied:
I wouldn't assume that we check one direction before the other. That's why I would recommend sticking with "other direction".

antoninbas commented on pkg/agent/packetcapture/capture/bpf.go:
```diff
@@ -72,11 +72,74 @@ func compareProtocol(protocol uint32, skipTrue, skipFalse uint8) bpf.Instruction
+			inst = append(inst, loadIPv4DestinationPort)
+			inst = append(inst, bpf.JumpIf{Cond: bpf.JumpEqual, Val: uint32(dstPort), SkipTrue: 0, SkipFalse: size - curLen - uint8(len(inst)) - 2})
+		}
+
```
remove extra empty line

antoninbas commented on pkg/agent/packetcapture/capture/bpf.go:
```diff
@@ -169,50 +216,131 @@ func compilePacketFilter(packetSpec *crdv1alpha1.Packet, srcIP, dstIP net.IP) []
-			if transPort.TCP.DstPort != nil {
-				count += 2
+		transport := packet.TransportHeader
+		portFiltersSize := func() int {
```
the way you wrote it, there is no need to call the function twice I think. So maybe use it like this:
```go
portFiltersSize := func() int {
    // ...
}()
// ...
count += portFiltersSize
```

antoninbas commented on pkg/agent/packetcapture/capture/bpf_test.go:
```diff
@@ -173,11 +209,133 @@ func TestPacketCaptureCompileBPF(t *testing.T) {
+					Protocol: &testTCPProtocol,
+					TransportHeader: crdv1alpha1.TransportHeader{
+						TCP: &crdv1alpha1.TCPHeader{
+							SrcPort: &testSrcPort,
```
not introduced by you, but do you think you could change the test code so that `testSrcPort` and `testDstPort` use 2 different values? Maybe 12345 for the source and 80 for the destination.

AryanBakliwal replied:
Sure, I will change the code for tests.

antoninbas commented on pkg/agent/packetcapture/capture/bpf_test.go:
```diff
@@ -173,11 +209,133 @@ func TestPacketCaptureCompileBPF(t *testing.T) {
+				bpf.JumpIf{Cond: bpf.JumpBitsSet, Val: 0x1fff, SkipTrue: 6}, // do we have an L4 header?
+				bpf.LoadMemShift{Off: 14},                                   // calculate size of IP header
+				bpf.LoadIndirect{Off: 14, Size: 2},                          // src port
+				bpf.JumpIf{Cond: bpf.JumpEqual, Val: 0x50, SkipFalse: 3},    // port 23
```
again, this was copied and paster from existing code, but what does the comment "port 23" mean? The port value is not 23, is it referring to the jump destination (but there are less than 23 instructions...). @hangyan is that your comment?

hangyan replied:
yeah i think so. comments should be updated.

AryanBakliwal replied:
I will update the comments.

antoninbas reviewed (commented):
Some small comments but overall LGTM. Thanks for all the work and patience.

antoninbas commented on pkg/agent/packetcapture/capture/bpf.go:
```diff
@@ -72,11 +72,74 @@ func compareProtocol(protocol uint32, skipTrue, skipFalse uint8) bpf.Instruction
+	if skipRequestCheck {
+		inst = append(inst, bpf.JumpIf{Cond: bpf.JumpEqual, Val: srcAddrVal, SkipTrue: 0, SkipFalse: calculateSkipFalse(srcPort, dstPort)})
+	} else {
+		inst = append(inst, bpf.JumpIf{Cond: bpf.JumpEqual, Val: srcAddrVal, SkipTrue: 0, SkipFalse: size - curLen - uint8(len(inst)) - 2})
```
`len(inst)` is actually 0 here, so we could remove it?

or maybe we could define this function:
```go
// calculate skip size to jump to the final instruction (NO MATCH)
skipToEnd := func() uint8 {
        return size - curLen - uint8(len(inst)) - 2
}
```

then use `SkipFalse: skipToEnd()` everywhere in this function. I think it would make the code even more readable. What do you think?

AryanBakliwal replied:
Using this function is better, I will make this change.

antoninbas reviewed (approved):
LGTM, thanks for this great contribution and for your patience during the review process

@hangyan do you want to take another look?

hangyan commented:
> LGTM, thanks for this great contribution and for your patience during the review process
> 
> @hangyan do you want to take another look?

sure.  will do it now and give my feedback soon

hangyan commented:
LGTM.

Thanks @AryanBakliwal

AryanBakliwal commented:
Thank you @antoninbas and @hangyan for guiding me through this. Your support made it much easier :)

antoninbas commented:
/test-all

antoninbas commented:
@AryanBakliwal One if your new e2e tests has failed in CI. I'm copying the logs below for convenience. PTAL.

<details>
<summary>Logs for failed test</summary>

```
2025-03-06T09:55:13.0928191Z === CONT  TestPacketCapture/testPacketCaptureBasic/testPacketCaptureBasic/ipv4-tcp-both
2025-03-06T09:55:14.5665773Z     packetcapture_test.go:581: CR status not match, actual: {NumberCaptured:5 FilePath:sftp://10.96.206.31:22/upload/ipv4-tcp-both.pcapng Conditions:[{Type:PacketCaptureStarted Status:True LastTransitionTime:2025-03-06 09:55:13 +0000 UTC Reason:Started Message:} {Type:PacketCaptureComplete Status:True LastTransitionTime:2025-03-06 09:55:14 +0000 UTC Reason:Succeed Message:} {Type:PacketCaptureFileUploaded Status:True LastTransitionTime:2025-03-06 09:55:14 +0000 UTC Reason:Succeed Message:}]}, expected: {NumberCaptured:10 FilePath:sftp://10.96.206.31:22/upload/ipv4-tcp.pcapng Conditions:[{Type:PacketCaptureStarted Status:True LastTransitionTime:0001-01-01 00:00:00 +0000 UTC Reason:Started Message:} {Type:PacketCaptureComplete Status:True LastTransitionTime:0001-01-01 00:00:00 +0000 UTC Reason:Succeed Message:} {Type:PacketCaptureFileUploaded Status:True LastTransitionTime:0001-01-01 00:00:00 +0000 UTC Reason:Succeed Message:}]}
2025-03-06T09:55:14.6059211Z I0306 09:55:14.605566   26834 framework.go:2950] Copying file "/tmp/antrea/packetcapture/packets/ipv4-tcp-both.pcapng" from Pod kube-system/antrea-agent-9rqsg
2025-03-06T09:55:14.6060486Z     packetcapture_test.go:699: 
2025-03-06T09:55:14.6061597Z         	Error Trace:	/home/runner/work/antrea/antrea/test/e2e/packetcapture_test.go:699
2025-03-06T09:55:14.6063285Z         	            				/home/runner/work/antrea/antrea/test/e2e/packetcapture_test.go:598
2025-03-06T09:55:14.6065129Z         	            				/home/runner/work/antrea/antrea/test/e2e/packetcapture_test.go:473
2025-03-06T09:55:14.6066262Z         	Error:      	Not equal: 
2025-03-06T09:55:14.6067020Z         	            	expected: "10.244.2.52"
2025-03-06T09:55:14.6067893Z         	            	actual  : "10.244.0.115"
2025-03-06T09:55:14.6068216Z         	            	
2025-03-06T09:55:14.6068671Z         	            	Diff:
2025-03-06T09:55:14.6069069Z         	            	--- Expected
2025-03-06T09:55:14.6069508Z         	            	+++ Actual
2025-03-06T09:55:14.6069859Z         	            	@@ -1 +1 @@
2025-03-06T09:55:14.6070217Z         	            	-10.244.2.52
2025-03-06T09:55:14.6070573Z         	            	+10.244.0.115
2025-03-06T09:55:14.6071187Z         	Test:       	TestPacketCapture/testPacketCaptureBasic/testPacketCaptureBasic/ipv4-tcp-both
2025-03-06T09:55:14.6071826Z     packetcapture_test.go:700: 
2025-03-06T09:55:14.6072579Z         	Error Trace:	/home/runner/work/antrea/antrea/test/e2e/packetcapture_test.go:700
2025-03-06T09:55:14.6074019Z         	            				/home/runner/work/antrea/antrea/test/e2e/packetcapture_test.go:598
2025-03-06T09:55:14.6075207Z         	            				/home/runner/work/antrea/antrea/test/e2e/packetcapture_test.go:473
2025-03-06T09:55:14.6075904Z         	Error:      	Not equal: 
2025-03-06T09:55:14.6076481Z         	            	expected: "10.244.0.115"
2025-03-06T09:55:14.6077124Z         	            	actual  : "10.244.2.52"
2025-03-06T09:55:14.6077868Z         	            	
2025-03-06T09:55:14.6078287Z         	            	Diff:
2025-03-06T09:55:14.6078637Z         	            	--- Expected
2025-03-06T09:55:14.6079211Z         	            	+++ Actual
2025-03-06T09:55:14.6079690Z         	            	@@ -1 +1 @@
2025-03-06T09:55:14.6080143Z         	            	-10.244.0.115
2025-03-06T09:55:14.6080645Z         	            	+10.244.2.52
2025-03-06T09:55:14.6081466Z         	Test:       	TestPacketCapture/testPacketCaptureBasic/testPacketCaptureBasic/ipv4-tcp-both
2025-03-06T09:55:14.6081961Z     packetcapture_test.go:718: 
2025-03-06T09:55:14.6082648Z         	Error Trace:	/home/runner/work/antrea/antrea/test/e2e/packetcapture_test.go:718
2025-03-06T09:55:14.6083801Z         	            				/home/runner/work/antrea/antrea/test/e2e/packetcapture_test.go:598
2025-03-06T09:55:14.6084893Z         	            				/home/runner/work/antrea/antrea/test/e2e/packetcapture_test.go:473
2025-03-06T09:55:14.6085317Z         	Error:      	Not equal: 
2025-03-06T09:55:14.6085765Z         	            	expected: 80
2025-03-06T09:55:14.6086399Z         	            	actual  : 49834
2025-03-06T09:55:14.6087969Z         	Test:       	TestPacketCapture/testPacketCaptureBasic/testPacketCaptureBasic/ipv4-tcp-both
2025-03-06T09:55:14.6088858Z     packetcapture_test.go:699: 
2025-03-06T09:55:14.6089458Z         	Error Trace:	/home/runner/work/antrea/antrea/test/e2e/packetcapture_test.go:699
2025-03-06T09:55:14.6090442Z         	            				/home/runner/work/antrea/antrea/test/e2e/packetcapture_test.go:598
2025-03-06T09:55:14.6091540Z         	            				/home/runner/work/antrea/antrea/test/e2e/packetcapture_test.go:473
2025-03-06T09:55:14.6092163Z         	Error:      	Not equal: 
2025-03-06T09:55:14.6092598Z         	            	expected: "10.244.2.52"
2025-03-06T09:55:14.6093046Z         	            	actual  : "10.244.0.115"
2025-03-06T09:55:14.6093488Z         	            	
2025-03-06T09:55:14.6093773Z         	            	Diff:
2025-03-06T09:55:14.6094116Z         	            	--- Expected
2025-03-06T09:55:14.6094462Z         	            	+++ Actual
2025-03-06T09:55:14.6094907Z         	            	@@ -1 +1 @@
2025-03-06T09:55:14.6095256Z         	            	-10.244.2.52
2025-03-06T09:55:14.6095599Z         	            	+10.244.0.115
2025-03-06T09:55:14.6096205Z         	Test:       	TestPacketCapture/testPacketCaptureBasic/testPacketCaptureBasic/ipv4-tcp-both
2025-03-06T09:55:14.6096819Z     packetcapture_test.go:700: 
2025-03-06T09:55:14.6097693Z         	Error Trace:	/home/runner/work/antrea/antrea/test/e2e/packetcapture_test.go:700
2025-03-06T09:55:14.6099330Z         	            				/home/runner/work/antrea/antrea/test/e2e/packetcapture_test.go:598
2025-03-06T09:55:14.6101419Z         	            				/home/runner/work/antrea/antrea/test/e2e/packetcapture_test.go:473
2025-03-06T09:55:14.6102254Z         	Error:      	Not equal: 
2025-03-06T09:55:14.6102985Z         	            	expected: "10.244.0.115"
2025-03-06T09:55:14.6103580Z         	            	actual  : "10.244.2.52"
2025-03-06T09:55:14.6103895Z         	            	
2025-03-06T09:55:14.6104310Z         	            	Diff:
2025-03-06T09:55:14.6104693Z         	            	--- Expected
2025-03-06T09:55:14.6105254Z         	            	+++ Actual
2025-03-06T09:55:14.6105776Z         	            	@@ -1 +1 @@
2025-03-06T09:55:14.6106253Z         	            	-10.244.0.115
2025-03-06T09:55:14.6106793Z         	            	+10.244.2.52
2025-03-06T09:55:14.6108147Z         	Test:       	TestPacketCapture/testPacketCaptureBasic/testPacketCaptureBasic/ipv4-tcp-both
2025-03-06T09:55:14.6109011Z     packetcapture_test.go:718: 
2025-03-06T09:55:14.6110100Z         	Error Trace:	/home/runner/work/antrea/antrea/test/e2e/packetcapture_test.go:718
2025-03-06T09:55:14.6111940Z         	            				/home/runner/work/antrea/antrea/test/e2e/packetcapture_test.go:598
2025-03-06T09:55:14.6113729Z         	            				/home/runner/work/antrea/antrea/test/e2e/packetcapture_test.go:473
2025-03-06T09:55:14.6114483Z         	Error:      	Not equal: 
2025-03-06T09:55:14.6115090Z         	            	expected: 80
2025-03-06T09:55:14.6115957Z         	            	actual  : 49834
2025-03-06T09:55:14.6116999Z         	Test:       	TestPacketCapture/testPacketCaptureBasic/testPacketCaptureBasic/ipv4-tcp-both
2025-03-06T09:55:14.6119002Z     packetcapture_test.go:598: 
2025-03-06T09:55:14.6120589Z         	Error Trace:	/home/runner/work/antrea/antrea/test/e2e/packetcapture_test.go:598
2025-03-06T09:55:14.6122658Z         	            				/home/runner/work/antrea/antrea/test/e2e/packetcapture_test.go:473
2025-03-06T09:55:14.6123398Z         	Error:      	Received unexpected error:
2025-03-06T09:55:14.6124333Z         	            	EOF
2025-03-06T09:55:14.6126047Z         	Test:       	TestPacketCapture/testPacketCaptureBasic/testPacketCaptureBasic/ipv4-tcp-both
2025-03-06T09:55:14.6127161Z === CONT  TestPacketCapture/testPacketCaptureBasic/testPacketCaptureBasic/ipv4-tcp
2025-03-06T09:55:16.1084962Z I0306 09:55:16.108202   26834 framework.go:2950] Copying file "/tmp/antrea/packetcapture/packets/ipv4-tcp.pcapng" from Pod kube-system/antrea-agent-9rqsg
2025-03-06T09:55:16.1149248Z === CONT  TestPacketCapture/testPacketCaptureBasic/testPacketCaptureBasic/non-existing-pod
2025-03-06T09:55:17.1262610Z === CONT  TestPacketCapture/testPacketCaptureBasic/testPacketCaptureBasic/invalid-host-public-key
2025-03-06T09:55:23.4110071Z I0306 09:55:23.410656   26834 framework.go:2950] Copying file "/tmp/antrea/packetcapture/packets/ipv4-icmp.pcapng" from Pod kube-system/antrea-agent-9rqsg
2025-03-06T09:55:28.4360245Z I0306 09:55:28.435537   26834 framework.go:2950] Copying file "/tmp/antrea/packetcapture/packets/ipv4-icmp-timeout.pcapng" from Pod kube-system/antrea-agent-9rqsg
2025-03-06T09:55:34.6221684Z I0306 09:55:34.621816   26834 framework.go:2950] Copying file "/tmp/antrea/packetcapture/packets/ipv4-udp.pcapng" from Pod kube-system/antrea-agent-9rqsg
2025-03-06T09:55:39.4440890Z I0306 09:55:39.443728   26834 framework.go:2950] Copying file "/tmp/antrea/packetcapture/packets/invalid-host-public-key.pcapng" from Pod kube-system/antrea-agent-9rqsg
2025-03-06T09:55:52.4711478Z === NAME  TestPacketCapture
2025-03-06T09:55:52.4713224Z     fixtures.go:352: Exporting test logs to '/home/runner/work/antrea/antrea/log/TestPacketCapture/beforeTeardown.Mar06-09-55-52'
2025-03-06T09:55:55.2982465Z     fixtures.go:523: Deleting 'testpacketcapture-1m63oz7o' K8s Namespace
2025-03-06T09:55:55.3026064Z I0306 09:55:55.302338   26834 framework.go:882] Deleting Namespace testpacketcapture-1m63oz7o took 4.302301ms
2025-03-06T09:55:55.3031550Z --- FAIL: TestPacketCapture (52.69s)
2025-03-06T09:55:55.3033334Z     --- FAIL: TestPacketCapture/testPacketCaptureBasic (42.40s)
2025-03-06T09:55:55.3034117Z         --- FAIL: TestPacketCapture/testPacketCaptureBasic/testPacketCaptureBasic (0.00s)
2025-03-06T09:55:55.3035120Z             --- FAIL: TestPacketCapture/testPacketCaptureBasic/testPacketCaptureBasic/ipv4-tcp-both (1.52s)
2025-03-06T09:55:55.3036220Z             --- PASS: TestPacketCapture/testPacketCaptureBasic/testPacketCaptureBasic/ipv4-tcp (1.50s)
2025-03-06T09:55:55.3038200Z             --- PASS: TestPacketCapture/testPacketCaptureBasic/testPacketCaptureBasic/non-existing-pod (1.01s)
2025-03-06T09:55:55.3039872Z             --- PASS: TestPacketCapture/testPacketCaptureBasic/testPacketCaptureBasic/ipv4-icmp (10.32s)
2025-03-06T09:55:55.3041535Z             --- PASS: TestPacketCapture/testPacketCaptureBasic/testPacketCaptureBasic/ipv4-icmp-timeout (15.35s)
2025-03-06T09:55:55.3043277Z             --- PASS: TestPacketCapture/testPacketCaptureBasic/testPacketCaptureBasic/ipv4-udp (21.54s)
2025-03-06T09:55:55.3045166Z             --- PASS: TestPacketCapture/testPacketCaptureBasic/testPacketCaptureBasic/invalid-host-public-key (22.32s)
```

</details>

AryanBakliwal commented:
/test-all

AryanBakliwal commented:
/test-e2e

AryanBakliwal commented:
@antoninbas the e2e test is passing now

antoninbas commented:
/test-all

antoninbas commented on test/e2e/packetcapture_test.go:
```diff
@@ -649,8 +696,13 @@ func verifyPacketFile(t *testing.T, pc *crdv1alpha1.PacketCapture, reader io.Rea
+		if pc.Spec.Direction == crdv1alpha1.CaptureDirectionBoth {
+			assert.Contains(t, []string{srcIP.String(), dstIP.String()}, ip.SrcIP.String())
+			assert.Contains(t, []string{srcIP.String(), dstIP.String()}, ip.DstIP.String())
+		} else {
```
this code would be wrong for the `DestinationToSource` direction. Could you add a single e2e test case for the `DestinationToSource` direction as a follow-up PR, and update this verification function accordingly?

AryanBakliwal replied:
Sure, will add it

antoninbas commented on test/e2e/packetcapture_test.go:
```diff
@@ -667,11 +719,20 @@ func verifyPacketFile(t *testing.T, pc *crdv1alpha1.PacketCapture, reader io.Rea
+					if ports.SrcPort != nil {
+						assert.Contains(t, []int32{int32(tcp.SrcPort), int32(tcp.DstPort)}, *ports.SrcPort)
+					}
+				} else {
```
ditto
